mlflow/pyfunc/__init__.py

```diff
@@ -656,12 +675,154 @@ def _warn_dependency_requirement_mismatches(model_path):
+            # the data stored in the disk is shared with all remote nodes.
+            root_dir = os.path.join(nfs_root_dir, "venvs")
+            os.makedirs(root_dir, exist_ok=True)
+            _VIRTUAL_ENV_DIR = tempfile.mkdtemp(dir=root_dir)
+            # TODO: register deleting _VIRTUAL_ENV_DIR handler when exit
+        else:
+            import atexit
+            import shutil
+            _VIRTUAL_ENV_DIR = tempfile.mkdtemp()
+            atexit.register(shutil.rmtree, _VIRTUAL_ENV_DIR, ignore_errors=True)
+
+    return _VIRTUAL_ENV_DIR
+
+
+_PICKLE_PROTOCOL_FOR_RESTORE_PY_ENV = 3
+
+
+def _setup_restored_python_env(py_version, requirements_file_path):
+    import shutil
+    # TODO: support py_version format (major.minor), and install latest micro version
+    pyenv_home_path = _get_or_create_pyenv_home_dir()
+
+    # install specified python version
+    pyenv_proc_env = os.environ.copy()
+    pyenv_proc_env['HOME'] = pyenv_home_path
+    subprocess.run(
+        [_PYENV_CMD, 'install', '-v', '-s', py_version],
+        check=True, stdout=sys.stdout, stderr=sys.stderr,
+        env=pyenv_proc_env
+    )
+
+    python_bin_path = os.path.join(
+        pyenv_home_path, f'.pyenv/versions/{py_version}/bin/python'
+    )
+
+    venv_cache_root_dir = _get_or_create_python_virtual_env_temp_dir()
+
+    req_file_content_sha = get_file_sha_hexdigest(requirements_file_path)
+    virtual_env_dir = os.path.join(
+        venv_cache_root_dir, req_file_content_sha
+    )
+    if not os.path.exists(virtual_env_dir):
+        try:
+            # install virtual env
+            subprocess.run(
+                [_VIRTUALENV_CMD, f'--python={python_bin_path}', virtual_env_dir]
+            )
+            virtual_env_pip = os.path.join(
+                virtual_env_dir, 'bin/pip'
+            )
+            subprocess.run(
+                [virtual_env_pip, 'install', '-r', requirements_file_path],
+                check=True, stdout=sys.stdout, stderr=sys.stderr
+            )
+            # TODO: log mlflow version when logging model and get mlflow version from logged model.
+            subprocess.run(
+                [virtual_env_pip, 'install', 'mlflow', 'gunicorn']
+            )
+        except Exception:
+            # If exception happens during setup virtual env, delete the broken virtual env directory.
+            # otherwise the broken virtual env is cached and follow-up setting up virtual env
+            # using the same requirement file will reusing the broken directory and failed again.
+            shutil.rmtree(virtual_env_dir, ignore_errors=True)
+            raise
+    else:
+        _logger.info(f'Found python virtual env cache at {virtual_env_dir}')
+
+    virtual_env_python_bin = os.path.join(
+        virtual_env_dir, 'bin/python'
+    )
+    return virtual_env_python_bin
+
+
+def _load_model_from_local_path(
+    local_path: str, suppress_warnings: bool = False
+):
```

Comment:
Do we need to make these changes? If the path already exists on the local FS, `download_artifact_from_uri` just returns it without doing any copying.

Reply:
In my test , No. `_download_artifact_from_uri` always creates a temporary directory and copy files into the directory.

Reply:
I don't think so.

I tried running the following code from my `/private/tmp` directory:

```
import mlflow

with mlflow.start_run():
    mlflow.log_dict({"a": "b"}, "mydict.json")
    rid = mlflow.active_run().info.run_id
    print(rid)


from mlflow.tracking.artifact_utils import _download_artifact_from_uri

print(_download_artifact_from_uri("runs:/" + rid + "/mydict.json"))
print(_download_artifact_from_uri("/private/tmp/mlruns/0/" + rid + "/artifacts/mydict.json"))
```

```
(base2) corey.zumar@C02F17H3MD6R /private/tmp $ python test.py
f27285ff6207401bb2f5b597c75c60ec
/private/tmp/mlruns/0/f27285ff6207401bb2f5b597c75c60ec/artifacts/mydict.json
/private/tmp/mlruns/0/f27285ff6207401bb2f5b597c75c60ec/artifacts/mydict.json
```

The original file path is being returned.

Reply:
got it . my fault. You are right.